File: new/is_valid_parentheses_stack.cpp
class Solution {
public:
    bool isValid(string s) {
        // create a stack to track the opening parentheses. 
        // create a map of closing parentheses => opening parentheses. 
        // loop through the given string, if the char not in map, add it to the stack cause it is opening
        // parentheses. 
        // if the char in map, check if the stack is not empty, and 
        // the map[ch] == stack's top (opening parentheses). if so pop the stack. 
        // if either of these cases are not true, return false. 
        // finally check if the stack is empty, if so return true, else return false. 
        stack<char> opening_parentheses_stack;
        unordered_map<char, char> closing_parentheses_to_opening_parentheses = {
            {')', '('}, 
            {']', '['}, 
            {'}', '{'}
        };
        for (char c: s) {
            if (closing_parentheses_to_opening_parentheses.count(c)) {
                if (!opening_parentheses_stack.empty() && closing_parentheses_to_opening_parentheses[c] == opening_parentheses_stack.top()) {
                    opening_parentheses_stack.pop();
                } else {
                    return false;
                }
            } else {
                opening_parentheses_stack.push(c);
            }
        }

        return opening_parentheses_stack.empty();
    }
    // time: O(n)
    // space: O(n)
};

File: new/python/is_valid_parentheses_stack.py
class Solution:
    def isValid(self, s: str) -> bool:
        # idea: create a map of closing parentheses to openeing parentheses. loop the string, if the char
        # not in the map, then it is not closing parentheses, it is opening parentheses, add it to the stack
        # of opening parentheses. if it is in the map, and the map, and stack not empty, and the char maps
        # to the one at the top of the stack, then we can pop the stack.
        # finally return true if the stack is empty, if not return false.
        stack = []  # contains opening parentheses.
        map = {
            ")": "(",
            "]": "[",
            "}": "{",
        }  # this is a map of closing parentheses -> opening parentheses.
        for ch in s:
            if ch in map:
                if stack and map[ch] == stack[-1]:
                    stack.pop()
                else:
                    return False  # if stack is empty, return False, or the closing bracket does not
                    # match the opening bracket in stack return False.
            else:
                stack.append(ch)

        return True if not stack else False
        # time: O(n) because we go through the string using for loop
        # space: O(n) because we use stack.
